src/DsaBottomUp.cc

```diff
@@ -50,10 +50,104 @@ static const Value *findUniqueReturnValue(const Function &F) {
   return onlyRetVal;
 }
 
+template <typename Set>
+static void markReachableNodes(const Node *n, Set &set) {
+  if (!n)
+    return;
+  assert(!n->isForwarding() && "Cannot mark a forwarded node");
+
+  if (set.insert(n).second)
+    for (auto const &edg : n->links())
+      markReachableNodes(edg.second->getNode(), set);
+}
+
+template <typename Set>
+static void reachableNodes(const Function &fn, Graph &g, Set &inputReach,
+                           Set &retReach) {
+  // formal parameters
+  for (const Value &arg : fn.args()) {
+    if (g.hasCell(arg)) {
+      Cell &c = g.mkCell(arg, Cell());
+      markReachableNodes(c.getNode(), inputReach);
+    }
+  }
+
+  // globals
+  for (auto &kv : g.globals())
+    markReachableNodes(kv.second->getNode(), inputReach);
+
+  // return value
+  if (g.hasRetCell(fn))
+    markReachableNodes(g.getRetCell(fn).getNode(), retReach);
+}
+
+// Copy (recursively) all callsites owned by calleeC to graph callerG.
+// Precondition: callerC simulates calleeC.
+static void copyCallSites(Graph &calleeG, const Cell &calleeC, Graph &callerG,
+                          Cell &callerC, std::set<Node *> &visited,
+                          std::set<DsaCallSite *> &copiedCallSites) {
```

Comment:
why `std::set`? Should be `std::vector`. Uniqueness is enforced by uniqueness of call sites in the source graph